flycheck.el

```diff
@@ -6920,6 +6921,62 @@ See URL `http://www.erlang.org/'."
    (error line-start (file-name) ":" line ": " (message) line-end))
   :modes erlang-mode)
 
+(defun contains-rebar-config (dir-name)
+  "Return DIR-NAME if DIR-NAME/rebar.config exists, nil otherwise."
+  (when (file-exists-p (expand-file-name "rebar.config" dir-name))
+    dir-name))
+
+(defun locate-rebar3-project-root (file-name &optional prev-file-name acc)
+  "Find the top-most rebar project root for source FILE-NAME.
+
+A project root directory is any directory containing a
+rebar.config file.  Find the top-most directory to move out of any
+nested dependencies.
+
+FILE-NAME is a source file for which to find the project.
+
+PREV-FILE-NAME helps us prevent infinite looping
+
+ACC is an accumulator that keeps the list of results, the first
+non-nil of which will be our project root.
+
+Return the absolute path to the directory"
+  (if (string= file-name prev-file-name)
+      (car (remove nil acc))
+    (let ((current-dir (file-name-directory file-name)))
+      (locate-rebar3-project-root
+       (directory-file-name current-dir)
+       file-name
+       (cons (contains-rebar-config current-dir) acc)))))
```

Comment:
I think it would help with maintainability if this function was based on `locate-dominating-file`. Something like this, maybe? (untested):

```
(defun locate-topmost-dominating-file (file name)
  (let ((result nil)
        (further-up file))
    (while further-up
      (setq further-up (locate-dominating-file further-up name))
      (when further-up
        (setq result further-up)
        (setq further-up (file-name-directory (directory-file-name further-up)))))
    result))
```